### PR TITLE
Fix #6311: autosummary: autosummary table gets confused by complex type hints

### LINKED ISSUE #6311: autosummary table gets confused by long / complex type hints
**Describe the bug**

While creating the table for `.. autosummary::`, the function

https://github.com/sphinx-doc/sphinx/blob/15bc5a32bb0b78b432803e55fffa10c801182c75/sphinx/ext/autosummary/__init__.py#L430

will not produce the expected results, some type hints remain, albeit broken up.  Here is the long function signature in question:

```py
from pathlib import Path
# ...
def filter_file(path: Union[Path, str], pattern: str,
                repl: Union[Callable[[Match], str], str], count: int = 0,
                flags: int = 0, backup_extension: str = ".orig",
                line_based: bool = False, demand_different: bool = True,
                encoding: Optional[str] = None) -> Path:
```

In the `.. autosummary::` table, the following is produced:

```
filter_file(path, str], pattern, repl, str], …)
```

The rendered docs can be found [online here](https://ci-exec.readthedocs.io/en/latest/api/patch.html#module-ci_exec.patch), but the root of the problem can be seen by the `path` and `repl` parameters (first and third parameter).

Stepping through, what actually happens is that everything just gets split on `', '`:

https://github.com/sphinx-doc/sphinx/blob/15bc5a32bb0b78b432803e55fffa10c801182c75/sphinx/ext/autosummary/__init__.py#L448-L454

So when we have something like `Union[Path, str]` or the (rather ugly) more complicated `Union[Callable[[Match], str], str]`, splitting on `', '` will give us something like this (splitting each argument on line to make clear):

```py
(Pdb) p args
[
    'path: Union[pathlib.Path',
    'str]',
    'pattern: str',
    'repl: Union[Callable[[Match[AnyStr]]',
    'str]',
    # ... remaining arguments ...
]
```

This is actually a pretty hard problem to solve with regular expressions, if it's even possible to do so (unless I'm missing something, it's kind of late xD).  I think we would have to start counting how many `[` we have seen and match them with `]` (after having seen a `:`).

So I was playing around with alternatives, found a nice SO answer that has some relevant stuff, here's a rough prototype:

```py
#!/usr/bin/env python3

import ast
import re
class HintDestroyer(ast.NodeTransformer):
    # See: https://stackoverflow.com/a/42734810/3814202
    # we only need to visit function since we only create function
    def visit_FunctionDef(self, node):
        node.returns = None
        if node.args.args:
            for arg in node.args.args:
                # just for a glimpse ...
                print("Removing annotation for {}: {}".format(arg.arg, arg.annotation.__dict__))
                arg.annotation = None
        return node

# sample input that breaks
sig = "(path: Union[pathlib.Path, str], pattern: str, repl: Union[Callable[[Match[AnyStr]], str], str], count: int = 0, flags: int = 0, backup_extension: str = '.orig', line_based: bool = False, demand_different: bool = True, encoding: Optional[str] = None) -> pathlib.Path"

# create a valid function signature to parse with ast module.
function = "def foo{sig}: pass\n".format(sig=sig)

# parse and transform
parsed_function = ast.parse(function)
transformed = HintDestroyer().visit(parsed_function)
# we only created a single "function", there should be a single module with a
# function as its body.
transformed_function = transformed.body[0]

# is this what we want?
args = ", ".join([item.arg for item in transformed_function.args.args])
print(args)
# output:
# path, pattern, repl, count, flags, backup_extension, line_based, demand_different, encoding
```

Right now I _believe_ what `args` and `opts` in the current autosummary code are doing is trying to shove `opts` at the end _if there is space_ (since there's a maximum length here).  I'm not sure why this happens, since all of the keyword arguments have to be last anyway right?

Basically we can use `ast` and just limited join on the transformed thing if we don't care about necessarily preserving existing behavior.  But I'm not sure about the implications of all this ... so thought I would open up the issue for discussion of other possible solutions or maybe downsides to use `ast` (it definitely doesn't come for free, but it's also a simple parse).

**To Reproduce**

```
$ git clone https://github.com/svenevs/ci_exec.git
$ cd ci_exec
$ tox -e docs
$ open .tox/docs/tmp/html/index.html

# ... or ...

$ pip install -r docs/requirements.txt
$ cd docs
$ make html 
$ open _build/html/index.html
```

**Expected behavior**
No stray `str]` end up as a "parameter" in the autosummary table.

**Your project**
https://github.com/svenevs/ci_exec

**Screenshots**

over_parse

**Environment info**
- OS: linux
- Python version: 3.7.2
- Sphinx version: 2.0.1
- Sphinx extensions:  sphinx.ext.autosummary
- Extra tools: n/a

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Bugfix

### Purpose
- refs: #6311

### CODE REVIEW

File: sphinx/ext/autosummary/__init__.py [hunk 1]
```diff
@@ -329,7 +329,12 @@ def get_items(self, names):
 
             # -- Grab the signature
 
-            sig = documenter.format_signature()
+            try:
+                sig = documenter.format_signature(show_annotation=False)
```
Comment (tk0miya) [inline sphinx/ext/autosummary/__init__.py:333]: But this handles `TypeError` to support custom documenters. So I believe this does not break compatibility.

**Comment by svenevs** [comment]
Just wanted to drop a quick note saying

1. Thank you!!!  This looks like a much better resolution.
2. I can test it either late tonight or tomorrow (mother's day here in the states :))

**Comment by svenevs** [comment]
Just tested it out, works perfectly!!!  Very cool :slightly_smiling_face:

**Comment by tk0miya** [comment]
Thank you for verifying!

**Comment by svenevs** [comment]
DISREGARD!  I got confused.

~There seem to have been bad merge?~

- ~[merge 2.0 into 6311_autosummary_confused_by_complex_typehints2](https://github.com/sphinx-doc/sphinx/commit/59d0f076ec2bd50ad0e59ce81e1fef1e4c1a9cd1)~
- ~[merge branch back to 2.0](https://github.com/sphinx-doc/sphinx/commit/774c59821c3c71c9aaac63f51d957264d30311ab)~

~Right now if I `pip install git+https://github.com/sphinx-doc/sphinx.git@2.0` things don't work again, but they worked when I tested before 2.0 was merged to 6311_... branch (now deleted).  Steps to replicate:~

```console
sven:~/Desktop> python -m venv venv
sven:~/Desktop> source venv/bin/activate
(venv) sven:~/Desktop> git clone https://github.com/svenevs/ci_exec.git
Cloning into 'ci_exec'...
remote: Enumerating objects: 126, done.
remote: Counting objects: 100% (126/126), done.
remote: Compressing objects: 100% (80/80), done.
remote: Total 253 (delta 76), reused 80 (delta 44), pack-reused 127
Receiving objects: 100% (253/253), 102.80 KiB | 1.09 MiB/s, done.
Resolving deltas: 100% (134/134), done.
(venv) sven:~/Desktop> cd ci_exec/docs/
(venv) sven:~/Desktop/ci_exec/docs (master)> pip install git+https://github.com/sphinx-doc/sphinx.git@2.0
Collecting git+https://github.com/sphinx-doc/sphinx.git@2.0
  Cloning https://github.com/sphinx-doc/sphinx.git (to revision 2.0) to /tmp/pip-req-build-i21gi6uc
   ...
(venv) sven:~/Desktop/ci_exec/docs (master)> pip install -r requirements.txt
...
(venv) sven:~/Desktop/ci_exec/docs (master)> make html
...
(venv) sven:~/Desktop/ci_exec/docs (master)> open build/html/api/patch.html
```

~Autosummary table at the top of that page is not doing what it did when I tested this :/  You can see similar problems on the autosummary at the top of other modules there too.~

merge_regression

~I will re-review the diffs again, I combed through it for about an hour but don't see what went wrong.  Will need fresh eyes...~

**Comment by svenevs** [comment]
OMG I'm so sorry @tk0miya as soon as I started tracing through `mangle_signature` I realized that the `[, count, ...]` is intended behavior, these are the optional parameters as denoted by the `[...]`.

Thanks again for fixing this, sorry for getting confused about this behavior!

**Comment by tk0miya** [comment]
:-)

**Pull request merged**